File: CSharpDotNetCore/AOC2020/Day04.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace AOC2020
{
    public static class Day04
    {
        public static int part1(string input) => parse(input).Count(isValid1);
        public static int part2(string input) => parse(input).Count(isValid2);

        private static IEnumerable<Dictionary<string, string>> parse(string input)
        {
            var result = new List<Dictionary<string, string>>();
            var next = new Dictionary<string, string>();

            foreach (var line in input.Split("\n"))
                if (string.IsNullOrEmpty(line))
                {
                    if (next.Count == 0) continue;
                    result.Add(next);
                    next = new Dictionary<string, string>();
                }
                else
                    foreach (var field in line.Split(" "))
                    {
                        var split = field.Split(":");
                        next.Add(split[0], split[1]);
                    }

            if (next.Count > 0) result.Add(next);

            return result;
        }

        private static bool isValid1(Dictionary<string,string> input) =>
            input.ContainsKey("cid")
                ? input.Count == 8
                : input.Count == 7;

        private static bool isValid2(Dictionary<string, string> input) => 
            isValid1(input) && input.All(validField);

        private static bool validField(KeyValuePair<string, string> field) =>
            field.Key switch
            {
                "byr" => isInt(field.Value, 1920, 2002),
                "iyr" => isInt(field.Value, 2010, 2020),
                "eyr" => isInt(field.Value, 2020, 2030),
                "hgt" => isHeight(field.Value),
                "hcl" => isHairColor(field.Value),
                "ecl" => isEyeColor(field.Value),
                "pid" => isId(field.Value),
                "cid" => true,
                _ => false
            };

        private static bool isInt(string value, int min, int max)
        {
            try
            {
                var year = int.Parse(value);
                return year >= min && year <= max;
            }
            catch
            {
                return false;
            }
        }

        private static bool isHeight(string value) =>
            value.EndsWith("cm")
                ? isInt(value.Replace("cm", ""), 150, 193)
                : value.EndsWith("in") && 
                  isInt(value.Replace("in", ""), 59, 76);

        private static bool isHairColor(string value) =>
            Regex.IsMatch(value, "^#([0-9]|[a-f]){6}$");

        static readonly HashSet<string> validEyeColors = 
            new HashSet<string> {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"};

        private static bool isEyeColor(string value) => 
            validEyeColors.Contains(value);

        private static bool isId(string value) =>
            Regex.IsMatch(value, "^[0-9]{9}$");
    }
}

File: JS/src/day04.js
module.exports.part1 = function (input) {
  let passports = parseInput(input);
  return passports.filter(isValid1).length;
};

module.exports.part2 = function (input) {
  let passports = parseInput(input);
  return passports.filter(isValid2).length;
};

function parseInput(input) {
  return input
    .replace(/\r/g, "")
    .split(/\n\n/)
    .map((s) => s.replace(/\n/g, " ").trim());
}

function isValid1(line) {
  return line.split(" ").length === (line.indexOf("cid") >= 0 ? 8 : 7);
}

function isValid2(line) {
  return isValid1(line) && line.split(" ").every(validField);
}

function validField(field) {
  const split = field.split(":");
  const key = split[0];
  const value = split[1];

  switch (key) {
    case "byr":
      return isInt(value, 1920, 2002);
    case "iyr":
      return isInt(value, 2010, 2020);
    case "eyr":
      return isInt(value, 2020, 2030);
    case "hgt":
      return isHeight(value);
    case "hcl":
      return isHairColor(value);
    case "ecl":
      return isEyeColor(value);
    case "pid":
      return isId(value);
    case "cid":
      return true;
  }
}

function isInt(value, min, max) {
  const parsed = parseInt(value);
  return !isNaN(parsed) && parsed >= min && parsed <= max;
}

function isHeight(value) {
  return value.endsWith("cm")
    ? isInt(value.replace("cm", ""), 150, 193)
    : value.endsWith("in") && isInt(value.replace("in", ""), 59, 76);
}

function isHairColor(value) {
  return /^#([0-9|[a-f]){6}$/.test(value);
}

const validEyeColors = {
  amb: true,
  blu: true,
  brn: true,
  gry: true,
  grn: true,
  hzl: true,
  oth: true,
};

function isEyeColor(value) {
  return validEyeColors[value];
}

function isId(value) {
  return /^[0-9]{9}$/.test(value);
}
